frontend/src/pages/ArtifactList.tsx

```diff
@@ -184,7 +192,72 @@ class ArtifactList extends Page<{}, PipelineListState> {
         } else {
           return compare ? 1 : -1;
         }
+      }));
+  }
+
+  /**
+   * Groups the incoming rows by name and type pushing all but the first row
+   * of each group to the expandedRows Map.
+   * @param rows
+   */
+  private groupRows(rows: Row[]): Row[] {
+    const flattenedRows = rows.reduce((map, r) => {
+      const stringKey = `${r.otherFields[2]}/${r.otherFields[0]}`;
+      const rows = map.get(stringKey);
+      if (rows) {
+        rows.push(r);
+      } else {
+        map.set(stringKey, [r]);
+      }
+      return map;
+    }, new Map<string, Row[]>());
+
+    const grouped: Row[] = [];
+    const expandedRows = new Map<number, Row[]>();
+    Array.from(flattenedRows.entries())
+      .sort((r1, r2) => r1[0] < r2[0] ? -1 : 1)
+      .forEach((r, index) => {
+        grouped.push(r[1][0]);
+        expandedRows.set(index, r[1].slice(1));
       });
+
+    this.setState({expandedRows});
+    return grouped;
+  }
+
+  /**
+   * Toggles the expansion state of a row
+   * @param index
+   */
+  private toggleRowExpand(index: number): void {
```

Comment:
Perhaps it doesn't need to happen in this PR, but I feel like the expand logic in this file should be moved to the `CustomTable` so that it doesn't need to be reimplemented for future list pages

Reply:
I was initially thinking the same thing. I modeled this after what was done in Pipelines. To change it, I think we'd need to lift the expansion state out of the `Row` interface and maintain it within the `CustomTable`. Since `Rows` are passed in as props to `CustomTable` currently, there's no good way of affecting their state without calling back to the component that provided them. It does seem like a good modification for a future PR.

Reply:
Yes, I agree. That sounds like a better solution. Okay to do it later